### ISSUE TITLE
Proposal to rename Timing constructs to Given, When, Then, Teardown

### ISSUE THREAD

**Issue by slinkydeveloper** [knative-sandbox/reconciler-test#51]
I think we should renaming the `Timing` constructs as following (quotes from https://martinfowler.com/bliki/GivenWhenThen.html):

* `Setup` -> `Given`: _The given part describes the state of the world before you begin the behavior you're specifying in this scenario. You can think of it as the pre-conditions to the test._
* `Requirement` -> `When`: _The when section is that behavior that you're specifying._
* `Assert` -> `Then`: _Finally the then section describes the changes you expect due to the specified behavior._
* No change to `Teardown`

Most of our tests in eventing actually follow this logic, a classic eventing test usually works like:

* `Given`: A bunch of resources are created
* `When`: One or more events are sent
* `Then`: Assert if events are received

For example, let's pick this test https://github.com/knative-sandbox/eventing-kafka/blob/master/test/e2e/kafka_source_test.go#L50:

* `Given`:
  1. New topic on Kafka
  2. Event recorder up and ready
  3. KafkaSource up and ready
* `When`: Publish message on Kafka
* `Then`: Assert event recorder received the event

**Comment by n3wscott** [comment]
We never use `Assert` directly, it is intended to be fronted by `f.Stable`, `f.Beta` or `f.Alpha`.

**Comment by n3wscott** [comment]
`Setup` is not `Given`, Given phases are intended to be parameters:

> The given part describes the state of the world before you begin the behavior you're specifying in this scenario. You can think of it as the pre-conditions to the test.

But `Setup` is more of actively adjust the environment for the following feature.

`Requirement` is not `When` as I have been thinking about this framework.  `Requirement` is asserting something about the test environment given the `Setup` phase has passed.

At first look, what you are proposing makes sense, but it leads to usage of the framework in a way I did not indent. I think it would be better to use a TDD framework that lets you compose tests in the Given When Then style inside an assertion. My thinking is this e2e/conformance test helps you orchestrate the environment and those moving pieces, plus filtering of assertion levels.

**Comment by slinkydeveloper** [comment]
> But Setup is more of actively adjust the environment for the following feature.

In my mind I'm mapping mapping scenario = feature, that's why I don't see the difference between Setup and When. Maybe it's just me that i didn't properly get it.

> Requirement is not When as I have been thinking about this framework. Requirement is asserting something about the test environment given the Setup phase has passed.

And if the requirement phase doesn't pass, what happens? Can you show me a use case where you use `Requirement` properly?

**Comment by n3wscott** [comment]
> mapping scenario = feature,
It could be I used the wrong word... I was trying to make feature == a check box in the spec report for an e2e or conformance test, but that is hard to distil. 

>And if the requirement phase doesn't pass, what happens? Can you show me a use case where you use Requirement properly?

I am developing these now in eventing, but at the moment if a Setup or Requirement does not pass, the `Test` fails.

I am assuming a Feature would have some requirement on "I have a ready resource that I did not create" or "FooKind is installed in the cluster", or even "K8s version 1.19 or larger is running"

**Comment by slinkydeveloper** [comment]
> I am assuming a Feature would have some requirement on "I have a ready resource that I did not create" or "FooKind is installed in the cluster", or even "K8s version 1.19 or larger is running"

Ok so requirement goes first than `Setup` in this sense. I guess if `Requirement` fails, the test should be ignored right?

So with your comments what I understood is:

* `Setup` is kinda Given + When
* `Requirement` is something you assert to figure out if you wanna proceed with testing the feature or not
* `Assert` is the Then part

Am I right?

**Comment by n3wscott** [comment]
Yeah that is close to how I have been thinking about this. Does it make sense to think of it like that or should we change something?

**Comment by slinkydeveloper** [comment]
I think now it's a little bit more clear to me, I guess we could do something like that to make stuff simpler/more explicit:

* `Requirement`
* `Given`
* `When`
* `Then`
* `Teardown`

But it's totally fine if we don't wanna do it, the whole problem in this issue was my misunderstanding of the timings.

**Comment by n3wscott** [comment]
I think this removes setup and explodes assert into Given, When, Then and there is nothing to stop you from using a testing lib that uses that structure. The framework has a goal of allowing the ordering and filtering of feature levels and states, not being a BDD framework

**Comment by slinkydeveloper** [comment]
I don't think you can explode the existing assert into given, when, then, because then there is an overlap between given and setup.

**Comment by slinkydeveloper** [comment]
> Requirement is not When as I have been thinking about this framework. Requirement is asserting something about the test environment given the Setup phase has passed.

I'm reopening this thread because, after looking more in the test framework, I still don't get the point of requirement. What is it meant for? And how is practically different from a Setup? Aren't our `Setup` at the end of the day already performing assertions (e.g. on the readyness of the system to test)?

**Comment by n3wscott** [comment]
>  I still don't get the point of requirement. What is it meant for? 

I was thinking it is for non-mutating phases. So it does not install or adjust anything, it is your chance to probe the system for things that are (or are not) installed or available as a gate to understand if the feature was to be run, would it be known to fail without trying (like it is a kafka feature but no kafka is installed) or it needs eventing but no eventing...

**Comment by slinkydeveloper** [comment]
But in `Setup` you already check if the system is ready or not, and we don't actually use requirement failures to skip tests today... So maybe it's a problem of how i see it implemented today?

**Comment by slinkydeveloper** [comment]
I think #106 solves this conversation, I'm closing this issue as not relevant anymore
